src/sentry/tasks/files.py

```diff
@@ -34,7 +35,15 @@ def delete_unreferenced_blobs(blob_ids):
         if FileBlobIndex.objects.filter(blob_id=blob_id).exists():
             continue
         try:
-            # Need to delete the record to ensure django hooks run.
-            FileBlob.objects.get(id=blob_id).delete()
+            blob = FileBlob.objects.get(id=blob_id)
         except FileBlob.DoesNotExist:
             pass
+        else:
+            try:
+                with transaction.atomic():
+                    # Need to delete the record to ensure django hooks run.
+                    blob.delete()
+            except IntegrityError:
+                # Do nothing if the blob was deleted in another task, or
+                # if had another reference added concurrently.
+                pass
```

Comment:
We have `on_delete=models.CASCADE` set for all FKs by default. I think that if another `FileBlobIndex` is created it might not error here, but will instead delete the `FileBlobIndex` row. Might be worth setting `on_delete=models.PROTECT` here and testing what happens if you try to delete a `FileBlob` with an existing relation.

Reply:
Wouldn't cascading by default mean we wouldn't be hitting integrity errors right now though?

Reply:
It depends on the timing. It will only hit `IntegrityError` if it tries to delete the blob and there's a `FileBlobIndex` row present. Django handles cascading deletions itself, it doesn't set up cascading deletions in the database. So the situation I imagine is:

`blob.delete()`: sub operations
 - check for related rows to cascade delete
 - if exists, perform `on_delete` action - This would cause any newly created rows to be deleted - this is actually bad, since we're incorrectly deleting rows
 - finally, delete the blob. If a new row was created between the previous check and this step, then that would cause the integrity error. I suspect that is how you're getting

If you take a look at the full stack trace in that error, you can see:
```
        collector = Collector(using=using)
        collector.collect([self], keep_parents=keep_parents)
        return collector.delete() # Error comes from here
```
My understanding is that `collector.collect` fetches all the things to cascade delete. So in the time between collection and deletion, I could definitely imagine new rows being created.

Reply:
So to make things safer here, it might be wise to explicitly specify `PROTECT` on that relationship. As for the integrity errors, I think what you have here is correct.

Reply:
Thanks for the extra context. I've modified the relationship definition.